// @ionic/core/dist/collection/utils/gesture/button-active.js
import { writeTask } from '@stencil/core';
import { hapticSelectionChanged, hapticSelectionEnd, hapticSelectionStart } from '../native/haptic';
import { createGesture } from './index';
export const createButtonActiveGesture = (el, isButton) => {
    let touchedButton;
    const activateButtonAtPoint = (x, y, hapticFeedbackFn) => {
        if (typeof document === 'undefined') {
            return;
        }
        const target = document.elementFromPoint(x, y);
        if (!target || !isButton(target)) {
            clearActiveButton();
            return;
        }
        if (target !== touchedButton) {
            clearActiveButton();
            setActiveButton(target, hapticFeedbackFn);
        }
    };
    const setActiveButton = (button, hapticFeedbackFn) => {
        touchedButton = button;
        const buttonToModify = touchedButton;
        writeTask(() => buttonToModify.classList.add('ion-activated'));
        hapticFeedbackFn();
    };
    const clearActiveButton = (dispatchClick = false) => {
        if (!touchedButton) {
            return;
        }
        const buttonToModify = touchedButton;
        writeTask(() => buttonToModify.classList.remove('ion-activated'));
        if (dispatchClick) {
            touchedButton.click();
        }
        touchedButton = undefined;
    };
    return createGesture({
        el,
        gestureName: 'buttonActiveDrag',
        threshold: 0,
        onStart: ev => activateButtonAtPoint(ev.currentX, ev.currentY, hapticSelectionStart),
        onMove: ev => activateButtonAtPoint(ev.currentX, ev.currentY, hapticSelectionChanged),
        onEnd: () => {
            clearActiveButton(true);
            hapticSelectionEnd();
        }
    });
};
